File: gke_customization/gke_catalog/report/employee_material_stock_in_hand/employee_material_stock_in_hand.py
# Copyright (c) 2026, Gurukrupa Export and contributors
# For license information, please see license.txt

import frappe
from frappe import _


def execute(filters=None):
    filters = filters or {}
    columns = get_columns()
    data    = get_summary_data(filters)
    return columns, data


def get_columns():
    return [
        {
            "label": _("Employee"),
            "fieldname": "employee",
            "fieldtype": "Link",
            "options": "Employee",
            "width": 120,
        },
        {
            "label": _("Employee Name"),
            "fieldname": "employee_name",
            "fieldtype": "Data",
            "width": 160,
        },
        {
            "label": _("Total Orders"),
            "fieldname": "total_orders",
            "fieldtype": "Int",
            "width": 110,
        },
        {
            "label": _("Dia Pcs"),
            "fieldname": "dia_pcs",
            "fieldtype": "Float",
            "width": 100,
            "precision": 2,
        },
        {
            "label": _("Dia Wt"),
            "fieldname": "dia_wt",
            "fieldtype": "Float",
            "width": 110,
            "precision": 3,
        },
        {
            "label": _("Stone Wt"),
            "fieldname": "stone_wt",
            "fieldtype": "Float",
            "width": 120,
            "precision": 3,
        },
        {
            "label": _("Gold Wt"),
            "fieldname": "gold_wt",
            "fieldtype": "Float",
            "width": 110,
            "precision": 3,
        },
        {
            "label": _("Details"),
            "fieldname": "details",
            "fieldtype": "HTML",
            "width": 100,
        },
    ]


# ── SUMMARY DATA (one row per employee) ───────────────────────────────────────

def get_summary_data(filters):
    conditions, values = build_conditions(filters)

    rows = frappe.db.sql(
        """
        WITH latest_txn AS (
            -- For every manufacturing_work_order, rank its Issue/Receive
            -- transactions by recency. Only the most recent one reflects
            -- who currently holds it (a work order is issued, received,
            -- re-issued, received again, etc. many times over its life).
            SELECT
                eo.parent                  AS ir_name,
                eo.manufacturing_work_order,
                eo.diamond_pcs,
                eo.diamond_wt,
                eo.gemstone_wt,
                eo.net_wt,
                ir.employee,
                ir.type,
                ir.department,
                ir.date_time,
                ir.docstatus,
                ROW_NUMBER() OVER (
                    PARTITION BY eo.manufacturing_work_order
                    ORDER BY ir.date_time DESC, ir.creation DESC
                ) AS rn
            FROM `tabEmployee IR Operation` eo
            JOIN `tabEmployee IR` ir
                ON eo.parent = ir.name
            WHERE
                ir.docstatus = 1
                AND eo.manufacturing_work_order IS NOT NULL
        )
        SELECT
            lt.employee                                AS employee,
            ee.employee_name                           AS employee_name,
            COUNT(DISTINCT lt.manufacturing_work_order) AS total_orders,
            COALESCE(SUM(lt.diamond_pcs), 0)            AS dia_pcs,
            COALESCE(SUM(lt.diamond_wt),  0)            AS dia_wt,
            COALESCE(SUM(lt.gemstone_wt), 0)            AS stone_wt,
            COALESCE(SUM(lt.net_wt),      0)            AS gold_wt
        FROM latest_txn lt
        JOIN `tabEmployee` ee
            ON ee.employee = lt.employee
        WHERE
            lt.rn = 1
            AND lt.type = 'Issue'
            {conditions}
        GROUP BY
            lt.employee, ee.employee_name
        HAVING
            COALESCE(SUM(lt.diamond_pcs), 0) != 0
            OR COALESCE(SUM(lt.diamond_wt), 0) != 0
            OR COALESCE(SUM(lt.gemstone_wt), 0) != 0
            OR COALESCE(SUM(lt.net_wt),      0) != 0
        ORDER BY
            ee.employee_name
        """.format(conditions=conditions),
        values,
        as_dict=True,
    )

    for row in rows:
        emp   = frappe.utils.escape_html(row["employee"])
        ename = frappe.utils.escape_html(row["employee_name"])
        row["details"] = (
            '<button'
            ' class="btn btn-xs btn-primary ir-detail-btn"'
            ' data-employee="{emp}"'
            ' data-empname="{ename}"'
            ' style="white-space:nowrap">'
            ' <i class="fa fa-list"></i> View'
            '</button>'
        ).format(emp=emp, ename=ename)

    return rows


# ── FILTERS → SQL CONDITIONS ──────────────────────────────────────────────────

def build_conditions(filters):
    conditions = []
    values = {}

    if filters.get("employee"):
        conditions.append("lt.employee = %(employee)s")
        values["employee"] = filters["employee"]

    if filters.get("department"):
        conditions.append("lt.department = %(department)s")
        values["department"] = filters["department"]

    if filters.get("from_date"):
        conditions.append("DATE(lt.date_time) >= %(from_date)s")
        values["from_date"] = filters["from_date"]

    if filters.get("to_date"):
        conditions.append("DATE(lt.date_time) <= %(to_date)s")
        values["to_date"] = filters["to_date"]

    if filters.get("status"):
        # Map label to docstatus integer: Draft=0, Submitted=1
        status_map = {"Draft": 0, "Submitted": 1}
        if filters["status"] in status_map:
            conditions.append("lt.docstatus = %(status)s")
            values["status"] = status_map[filters["status"]]

    cond_str = ("AND " + " AND ".join(conditions)) if conditions else ""
    return cond_str, values


# ── WHITELISTED API – called by JS detail dialog ──────────────────────────────

@frappe.whitelist()
def get_employee_ir_details(employee):
    """
    Returns row-wise breakdown for the given employee: exactly the
    manufacturing_work_order rows currently in their hands (i.e. whose
    most recent submitted transaction is an Issue to this employee).
    These are the same rows that feed the summary report row totals.
    Called from the JS detail dialog via frappe.call().
    """
    rows = frappe.db.sql(
        """
        WITH latest_txn AS (
            SELECT
                eo.parent                  AS ir_name,
                eo.manufacturing_work_order,
                eo.diamond_pcs,
                eo.diamond_wt,
                eo.gemstone_wt,
                eo.net_wt,
                ir.employee,
                ir.type,
                ir.date_time,
                ir.docstatus,
                ROW_NUMBER() OVER (
                    PARTITION BY eo.manufacturing_work_order
                    ORDER BY ir.date_time DESC, ir.creation DESC
                ) AS rn
            FROM `tabEmployee IR Operation` eo
            JOIN `tabEmployee IR` ir
                ON eo.parent = ir.name
            WHERE
                ir.docstatus = 1
                AND eo.manufacturing_work_order IS NOT NULL
        )
        SELECT
            lt.manufacturing_work_order AS Mfg_Work_Order,
            lt.ir_name                  AS ir_name,
            lt.date_time                AS posting_date,
            lt.docstatus                AS status,
            COALESCE(lt.diamond_pcs, 0) AS dia_pcs,
            COALESCE(lt.diamond_wt,  0) AS dia_wt,
            COALESCE(lt.gemstone_wt, 0) AS stone_wt,
            COALESCE(lt.net_wt,      0) AS gold_wt
        FROM latest_txn lt
        WHERE
            lt.rn = 1
            AND lt.type = 'Issue'
            AND lt.employee = %(employee)s
        ORDER BY
            lt.date_time DESC
        """,
        {"employee": employee},
        as_dict=True,
    )
    return rows

File: gke_customization/gke_catalog/report/employee_material_stock_in_hand/employee_material_stock_in_hand.js
// Copyright (c) 2026, Gurukrupa Export and contributors
// For license information, please see license.txt

frappe.query_reports["Employee Material Stock In Hand"] = {

    filters: [
        {
            fieldname: "from_date",
            label: __("From Date"),
            fieldtype: "Date",
            default: frappe.datetime.month_start(),
            reqd: 1,
        },
        {
            fieldname: "to_date",
            label: __("To Date"),
            fieldtype: "Date",
            default: frappe.datetime.get_today(),
            reqd: 1,
        },
        {
            fieldname: "employee",
            label: __("Employee"),
            fieldtype: "Link",
            options: "Employee",
        },
        {
            fieldname: "department",
            label: __("Department"),
            fieldtype: "Link",
            options: "Department",
        },
        {
            fieldname: "status",
            label: __("Status"),
            fieldtype: "Select",
            options: "\nDraft\nSubmitted",
            default: "",
        },
    ],

    onload: function (report) {
        // Remove any previously attached handler to avoid duplicates on refresh
        $(document).off("click.ir_detail_btn");

        // Use document-level delegation — survives DataTable re-renders
        $(document).on("click.ir_detail_btn", ".ir-detail-btn", function () {
            const employee = $(this).data("employee");
            const empname  = $(this).data("empname");
            frappe_show_ir_detail_dialog(employee, empname);
        });
    },
};


// Global so it can also be triggered via inline onclick as a fallback
window.frappe_show_ir_detail_dialog = function (employee, empname) {
    if (!employee) {
        frappe.msgprint(__("Employee not found on button."));
        return;
    }

    frappe.call({
        method: "gke_customization.gke_catalog.report.employee_material_stock_in_hand.employee_material_stock_in_hand.get_employee_ir_details",
        args: { employee: employee },
        freeze: true,
        freeze_message: __("Loading details…"),
        callback: function (r) {
            if (r.exc) {
                frappe.msgprint(__("Error fetching details. Check console for more info."));
                return;
            }

            if (!r.message || !r.message.length) {
                frappe.msgprint(__("No records found for {0}.", [empname || employee]));
                return;
            }

            // Map docstatus integer to label
            const status_label = { 0: "Draft", 1: "Submitted" };

            // Totals
            let totals = { dia_pcs: 0, dia_wt: 0, stone_wt: 0, gold_wt: 0 };
            r.message.forEach(row => {
                totals.dia_pcs  += flt(row.dia_pcs);
                totals.dia_wt   += flt(row.dia_wt);
                totals.stone_wt += flt(row.stone_wt);
                totals.gold_wt  += flt(row.gold_wt);
            });

            const detail_rows = r.message.map(row => `
                <tr>
                    <td><a href="/app/employee-ir/${row.ir_name}" target="_blank">${row.ir_name}</a></td>
                    <td>${row.posting_date || ""}</td>
                    <td>${status_label[row.status] || row.status}</td>
                    <td>${row.Mfg_Work_Order || ""}</td>
                    <td style="text-align:right">${flt(row.dia_pcs,  2)}</td>
                    <td style="text-align:right">${flt(row.dia_wt,   3)}</td>
                    <td style="text-align:right">${flt(row.stone_wt, 3)}</td>
                    <td style="text-align:right">${flt(row.gold_wt,  3)}</td>
                </tr>`).join("");

            const html = `
                <div style="overflow:auto; max-height:65vh;">
                    <table class="table table-bordered table-condensed" style="margin:0; font-size:12px;">
                        <thead style="position:sticky; top:0; background:#fff; z-index:1;">
                            <tr style="background:#f0f4f8;">
                                <th>IR ID</th>
                                <th>Date / Time</th>
                                <th>Status</th>
                                <th>Work Order</th>
                                <th style="text-align:right">Dia Pcs</th>
                                <th style="text-align:right">Dia Wt</th>
                                <th style="text-align:right">Stone Wt</th>
                                <th style="text-align:right">Net Wt</th>
                            </tr>
                        </thead>
                        <tbody>${detail_rows}</tbody>
                        <tfoot>
                            <tr style="font-weight:bold; background:#f5f5f5;">
                                <td colspan="4">Total &nbsp;(${r.message.length} record${r.message.length !== 1 ? "s" : ""})</td>
                                <td style="text-align:right">${flt(totals.dia_pcs,  2)}</td>
                                <td style="text-align:right">${flt(totals.dia_wt,   3)}</td>
                                <td style="text-align:right">${flt(totals.stone_wt, 3)}</td>
                                <td style="text-align:right">${flt(totals.gold_wt,  3)}</td>
                            </tr>
                        </tfoot>
                    </table>
                </div>`;

            const d = new frappe.ui.Dialog({
                title: __("IR Details — {0} ({1})", [empname || "", employee]),
                fields: [{ fieldtype: "HTML", fieldname: "detail_html" }],
                size: "extra-large",
            });
            d.fields_dict.detail_html.$wrapper.html(html);
            d.show();
        },
    });
};